### PR TITLE
Jamie/bicaws 1175 multiple group selection

### PR DESCRIPTION
**Pull request by bleakcon**

### CODE REVIEW

File: src/pages/users/[username]/edit.tsx
Comment (MihaiMadeT) [inline src/pages/users/[username]/edit.tsx:51]: (minor) you can move the groupsChecked initialisation and the initialisation of the groups from userDetails after the getUserById check? If we fail to get the user here, then the data we have slightly wasted computation to do the filtering and initialisation.

File: src/useCases/createUser.ts
Comment (MihaiMadeT) [inline src/useCases/createUser.ts:112]: can we return an error earlier if we do not have a currentUser.id? i.e. not have the SQL query return a failure in this case and handle this more elegantly?

File: src/useCases/getUserByUsername.ts
Comment (MihaiMadeT) [inline src/useCases/getUserByUsername.ts:19]: this does not look correct?

File: src/useCases/getUserByUsername.ts
Comment (MihaiMadeT) [inline src/useCases/getUserByUsername.ts:38]: we have already checked if user is null at line 31, no?

File: src/useCases/index.ts
Comment (MihaiMadeT) [inline src/useCases/index.ts:8]: comment out/cleanup?

File: test/useCases/getUserById.integration.test.ts
Comment (MihaiMadeT) [inline test/useCases/getUserById.integration.test.ts:126]: instead, can we do expect(actualUser.groups).toNotBet(undefined) / or null?
would be quicker to debug if the test fails. otherwise in the terminal you will only see fail, which is not very informative.

File: src/useCases/getUserByUsername.ts
Comment (bleakcon) [reply src/useCases/getUserByUsername.ts:19]: In what way?

File: src/useCases/getUserByUsername.ts
Comment (bleakcon) [reply src/useCases/getUserByUsername.ts:38]: It seems that way doesn't it

Comment (bleakcon) [reply src/useCases/getUserByUsername.ts:38]: Commit comment : 1

File: src/useCases/index.ts
Comment (bleakcon) [reply src/useCases/index.ts:8]: Review comment 2

File: test/useCases/getUserById.integration.test.ts
Comment (bleakcon) [reply test/useCases/getUserById.integration.test.ts:126]: It is because the actualUser is typed as a partial user and groups is optional, therefore it will get raised as a possible null reference which is why the if there. 

I don't think 'toNotBe' exists as a matcher?

File: src/useCases/getUserByUsername.ts
Comment (MihaiMadeT) [reply src/useCases/getUserByUsername.ts:19]: check the line below it :)

File: test/useCases/getUserById.integration.test.ts
Comment (MihaiMadeT) [reply test/useCases/getUserById.integration.test.ts:126]: You are right ... I think it is something like .not.toBe or something

File: src/pages/users/[username]/edit.tsx
Comment (bleakcon) [reply src/pages/users/[username]/edit.tsx:51]: Review comment 3

File: src/useCases/getUserByUsername.ts
Comment (bleakcon) [reply src/useCases/getUserByUsername.ts:19]: Hahah ok, probably an odd merge

Comment (bleakcon) [reply src/useCases/getUserByUsername.ts]: Review comment 4

File: test/useCases/getUserById.integration.test.ts
Comment (bleakcon) [reply test/useCases/getUserById.integration.test.ts:126]: `expect(actualUser.groups).not.toBeUndefined()` seems to do the trick in the else, I was hoping this might remove the need for an if/else but not so.

Comment (bleakcon) [reply test/useCases/getUserById.integration.test.ts:126]: Review comment 5

File: src/useCases/createUser.ts
Comment (bleakcon) [reply src/useCases/createUser.ts]: Review comment 6

File: test/useCases/getUserById.integration.test.ts
Comment (MihaiMadeT) [reply test/useCases/getUserById.integration.test.ts:126]: is it because it is actually a null maybe?

File: test/useCases/signInUser.integration.test.ts
Comment (MihaiMadeT) [inline test/useCases/signInUser.integration.test.ts:53]: you might want to remove exclusionList and inclusionList. Those will be removed at one point. They have been replaced in functionality by visibleForces, visibleCourts and excludedTriggers

**Review by daviesjamie — COMMENTED** [review]
Very tidy changes, looks good 👌 

I've got a couple of minor suggestions, though... 🙈 

- I think it would be good to "tidy up" the group names when displaying them to the user. `B7Allocator_grp` doesn't mean as much to the police users that will be using the interface as `Allocator` would. We should only do this mapping for the checkbox labels though, we should still use the `B7Allocator_grp` string internally and in the database.

- If it's not too much faff, I think it would be helpful to see the groups the user is in on the "User Details" page - as currently the only way of seeing which groups a user is in is to actually go through to the edit page.

**Comment by MihaiMadeT** [comment]
> Very tidy changes, looks good 👌
> 
> I've got a couple of minor suggestions, though... 🙈
> 
> * I think it would be good to "tidy up" the group names when displaying them to the user. `B7Allocator_grp` doesn't mean as much to the police users that will be using the interface as `Allocator` would. We should only do this mapping for the checkbox labels though, we should still use the `B7Allocator_grp` string internally and in the database.
> * If it's not too much faff, I think it would be helpful to see the groups the user is in on the "User Details" page - as currently the only way of seeing which groups a user is in is to actually go through to the edit page.
> 
> 

^ to add to this. In the groups table in postgres, there is a description column. We could use that and see that column when seeding the data. @Sam, would you like to pair on that to see how changes to postgres are done?

**Comment by bleakcon** [comment]
@MihaiMadeT  @daviesjamie 

Some good observations, I wonder though, whether the changes to the user details view should be in its own ticket as that seems to be outside the scope of this ticket.

We can pair on these changes but given this ticket is holding up other tickets being merged and these changes are not strictly part of this ticket what do you think to creating a few new bolt-on tickets that will have their own PRs for these changes?

**Comment by daviesjamie** [comment]
> ^ to add to this. In the groups table in postgres, there is a description column. We could use that and see that column when seeding the data and use the description column to make the groups more readable. @sam, would you like to pair on that to see how changes to postgres are done?

This is a good idea, but I think we should maybe create another ticket for it as it could turn into quite a bit more work and it arguably isn't a "must". What do you think?

When I said to "tidy up" the labels, I was thinking we could do a quick string transformation - something like:

```ts
// label = "B7GeneralHandler_grp"

label = label.replace(/^B7/, "")
// label = "GeneralHandler_grp"

label = label.replace(/_grp$/, "")
// label = "GeneralHandler"

label = label.replace(/([a-z])([A-Z])/, "$1 $2")
// label = "General Handler"
```

**Comment by MihaiMadeT** [comment]
> @MihaiMadeT @daviesjamie
> 
> Some good observations, I wonder though, whether the changes to the user details view should be in its own ticket as that seems to be outside the scope of this ticket.
> 
> We can pair on these changes but given this ticket is holding up other tickets being merged and these changes are not strictly part of this ticket what do you think to creating a few new bolt-on tickets that will have their own PRs for these changes?

I am ok with these going in and the fixes can be done as part of 2 more PRs. The changing of the names to something more legible involves another PR on the bichard-next repo

**Comment by MihaiMadeT** [comment]
> > ^ to add to this. In the groups table in postgres, there is a description column. We could use that and see that column when seeding the data and use the description column to make the groups more readable. @sam, would you like to pair on that to see how changes to postgres are done?
> 
> This is a good idea, but I think we should maybe create another ticket for it as it could turn into quite a bit more work and it arguably isn't a "must". What do you think?
> 
> When I said to "tidy up" the labels, I was thinking we could do a quick string transformation - something like:
> 
> ```ts
> // label = "B7GeneralHandler_grp"
> 
> label = label.replace(/^B7/, "")
> // label = "GeneralHandler_grp"
> 
> label = label.replace(/_grp$/, "")
> // label = "GeneralHandler"
> 
> label = label.replace(/([a-z])([A-Z])/, "$1 $2")
> // label = "General Handler"
> ```

Isn't that ... a bit hacky? What if we change the table row names at one point?

**Comment by bleakcon** [comment]
I think my approach to this would be:

1. whilst your suggestions are good and we should do some of them they are not a must, probably not part of the initial ticket?
2. I would like to add tickets for each of your suggestions and get them prioritised
3. I don't think this PR has anything that needs to be 'fixed', only additions, these feel outside scope or at least not part of the original ask.

My suggestion would be to get this merged as a priority and then work on these additional suggestions in isolation

**Review by daviesjamie — DISMISSED** [review]
Sounds good to me 👍

**Review by MihaiMadeT — APPROVED** [review]

**Review by MihaiMadeT — APPROVED** [review]
have another approval

File: test/useCases/signInUser.integration.test.ts
Comment (bleakcon) [reply test/useCases/signInUser.integration.test.ts]: Review comment 7

**Comment by bjpirt** [comment]
Agree with making two more tickets

**Pull request merged**
